### PR TITLE
Add new options for switch case statement blocks

### LINKED ISSUE #12304: Setting `Indent case contents` gives undesired behavior when braces are used to surround case contents.
If I set 'Indent case contents' in the C# options, things look fine when I don't use curly braces to surround the case contents:

```
switch (value)
{
    case ExampleEnum.Value1:
        // Do some stuff
        break;
}
```

But if I put that code in curly braces, it indents the braces as well (this doesn't look good IMO):

```
switch (value)
{
    case ExampleEnum.Value1:
        {
            // Do some stuff
            break;
        }
}
```

I think it is reasonable to say (when braces are present) the braces should be aligned with the case statement (or an option to do such should be provided, so-as to maintain back-compat):

```
switch (value)
{
    case ExampleEnum.Value1:
    {
        // Do some stuff
        break;
    }
}
```

### PR DESCRIPTION
**Pull request by viceroypenguin**
Change "Indent case contents" to be a radio button for three options:
1. Always indent case block
2. Never indent case block
3. Only indent if case block is statements and not if case block is a block statement.

Fixes #12304.

### CODE REVIEW

File: src/VisualStudio/CSharp/Impl/CSharpVSResources.resx [hunk 1]
```diff
@@ -489,4 +486,16 @@
   <data name="Show_completion_list_after_a_character_is_deleted" xml:space="preserve">
     <value>Show completion list after a character is _deleted</value>
   </data>
+  <data name="Always_indent_switch_case" xml:space="preserve">
```
Comment (tannergooding) [inline src/VisualStudio/CSharp/Impl/CSharpVSResources.resx]: Tagging @jinujoseph.

I think modifying the resource names may impact localization.

File: src/VisualStudio/CSharp/Impl/Options/AutomationObject.cs [hunk 2]
```diff
@@ -159,8 +159,15 @@ public int Indent_Braces
 
         public int Indent_CaseContents
         {
-            get { return GetBooleanOption(CSharpFormattingOptions.IndentSwitchCaseSection); }
-            set { SetBooleanOption(CSharpFormattingOptions.IndentSwitchCaseSection, value); }
+            get
+            {
+                return (int)_workspace.Options.GetOption(CSharpFormattingOptions.IndentSwitchCaseSection);
```
Comment (tannergooding) [inline src/VisualStudio/CSharp/Impl/Options/AutomationObject.cs]: Why no longer using the helper methods?

Comment (tannergooding) [inline src/VisualStudio/CSharp/Impl/Options/AutomationObject.cs]: Do we need to create a new option for back-compat?

I believe if the value was somehow set to `2` we would treat it as `AlwaysIndent` and now we will treat it as `OnlyIndentStatements`.

File: src/Workspaces/CSharp/Portable/Formatting/CSharpFormattingOptions.cs [hunk 3]
```diff
@@ -65,7 +65,7 @@ public static class CSharpFormattingOptions
 
         public static Option<bool> IndentSwitchSection { get; } = new Option<bool>(IndentFeatureName, "IndentSwitchSection", defaultValue: true);
 
-        public static Option<bool> IndentSwitchCaseSection { get; } = new Option<bool>(IndentFeatureName, "IndentSwitchCaseSection", defaultValue: true);
+        public static Option<SwitchCaseIndentOptions> IndentSwitchCaseSection { get; } = new Option<SwitchCaseIndentOptions>(IndentFeatureName, "IndentSwitchCaseSection", defaultValue: SwitchCaseIndentOptions.AlwaysIndent);
```
Comment (tannergooding) [inline src/Workspaces/CSharp/Portable/Formatting/CSharpFormattingOptions.cs]: This looks like a public API change. I think we need a new option to track the new functionality and map the old functionality into the new functionality.

**Comment by tannergooding** [comment]
Explicitly tagging @dotnet/roslyn-ide.

I'm personally in favor of this change. However, I think there needs to be some adjustments to maintain our existing public API surface area. Someone on the IDE team should be able to provide more specific details on what is required.

File: src/Workspaces/CSharp/Portable/PublicAPI.Shipped.txt [hunk 4]
```diff
@@ -9,7 +9,6 @@ Microsoft.CodeAnalysis.CSharp.Formatting.LabelPositionOptions.NoIndent = 2 -> Mi
 Microsoft.CodeAnalysis.CSharp.Formatting.LabelPositionOptions.OneLess = 1 -> Microsoft.CodeAnalysis.CSharp.Formatting.LabelPositionOptions
 static Microsoft.CodeAnalysis.CSharp.Formatting.CSharpFormattingOptions.IndentBlock.get -> Microsoft.CodeAnalysis.Options.Option<bool>
 static Microsoft.CodeAnalysis.CSharp.Formatting.CSharpFormattingOptions.IndentBraces.get -> Microsoft.CodeAnalysis.Options.Option<bool>
-static Microsoft.CodeAnalysis.CSharp.Formatting.CSharpFormattingOptions.IndentSwitchCaseSection.get -> Microsoft.CodeAnalysis.Options.Option<bool>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/PublicAPI.Shipped.txt]: This is a breaking change.

**Comment by CyrusNajmabadi** [comment]
This option seems pretty esoteric.  I'm not sure it's warranted given how long the product has been around without this.  Are there large existing C# codebases that use this style?

**Comment by tannergooding** [comment]
@CyrusNajmabadi, I believe so. Roslyn itself uses both case blocks that are only statements (https://github.com/dotnet/roslyn/blob/3466e42a4cc1c74bd00f2d1d9b7e5a60356621de/src/Compilers/CSharp/Portable/BoundTree/BoundStatementExtensions.cs#L16) and case blocks that are block statements (https://github.com/dotnet/roslyn/blob/3466e42a4cc1c74bd00f2d1d9b7e5a60356621de/src/Compilers/CSharp/Portable/FlowAnalysis/ControlFlowPass.cs#L189).

The first option (currently supported) does the following:
```C#
switch (value)
{
    case 0:
        // Some Code
        break;

    case 1:
        {
            // Some Code
            break;
        }
}
```

The second option (currently supported) does the following:
```C#
switch (value)
{
    case 0:
    // Some Code
    break;

    case 1:
    {
        // Some Code
        break;
    }
}
```

The third option (new) does the following
```C#
switch (value)
{
    case 0:
        // Some Code
        break;

    case 1:
    {
        // Some Code
        break;
    }
}
```

**Comment by CyrusNajmabadi** [comment]
I agree that people use statements and people use blocks within switches :)

What i'm less sure about is how necessary it is to have an option for people who want that specific 3rd style you've presented.

**Comment by tannergooding** [comment]
It's also probably going to be hard to find some existing code that matches the last option. Mainly because, even if it was written that way, the formatter would adjust it to match either option 1 or option 2.

**Comment by CyrusNajmabadi** [comment]
Right.

File: src/VisualStudio/CSharp/Impl/Options/AutomationObject.cs (hunk 2 above)
Comment (viceroypenguin) [inline src/VisualStudio/CSharp/Impl/Options/AutomationObject.cs]: I agree that there is a risk if the value was somehow set to two. Not sure how to get around that without setting the new option to a ridiculous value like 1000 or -1.  I did ensure that the existing defined values of 0 and 1 map to the same values going forward.

File: src/Workspaces/CSharp/Portable/Formatting/CSharpFormattingOptions.cs (hunk 3 above)
Comment (viceroypenguin) [inline src/Workspaces/CSharp/Portable/Formatting/CSharpFormattingOptions.cs]: I can do it that way if we want.  Would probably make sense given the breaking change comment below. I'll push another change today that will keep the original option and map to the new one.

File: src/VisualStudio/CSharp/Impl/Options/AutomationObject.cs (hunk 2 above)
Comment (viceroypenguin) [inline src/VisualStudio/CSharp/Impl/Options/AutomationObject.cs]: If I rewrite to keep the original option the same and add a new option for the tri-state, and map the original against the new, then this should resolve the concern here.

**Comment by viceroypenguin** [comment]
@CyrusNajmabadi - I would prefer option 3 myself.  Hence the interest in implementing... 😄

**Comment by viceroypenguin** [comment]
Added support for the old save files.  This doesn't fix the public API issue, but at least fixes the save file issue.

**Comment by CyrusNajmabadi** [comment]
> I would prefer option 3 myself. Hence the interest in implementing... :smile:

Obviously :smile:

Here's the issue for me though:  We recently did an investigation to see how some formatting options are actually used in practice.  We discovered, for example, that there were formatting options that were only changed by 0.0005% of our user base.  

To me, that's not serving the purpose of an option.  An option should be there when there's a reasonable expectation that a significant subsection of your users will use it.  Otherwise, that option is just adding costs to the codebase without providing enough value.  Indeed, with some of these options I actually believe there is negative value.  Primarily because it means that this 0.0005% of the userbase now has code that:

1. Looks differently than the code written by 99.9995% of the other developers out there.
2. Has code that will then be harder to edit/maintain by anyone using the default settings.
3. Runs a higher risk of regressing in our formatting engine because it is so rarely used that bugs in that area can go through without anyone noticing.

To me, the formatting engine is not a place for everyone to add their one-off tweaks to.  The formatting engine's purpose is to quickly and efficiently format code to the styles chosen by the large subsects of our users.  To that end, options like 'where do curly braces go?' definitely belong there.  As does handling of spaces/tabs/indentation levels.  I'd also like to see the formatting engine improve to handle things like "we like code to not go past 100 columns, but we'd like it to wrap sensibly when it does".  These are the sorts of features that end up justifying their cost.  Tiny little one-off options for a specific user, or handful of users, don't IMO.

**Comment by tannergooding** [comment]
@CyrusNajmabadi, if that is the case, then I believe the formatting engine should be pluggable.

In this manner, a user could write a formatting 'plugin' that either overrides the 'current' behavior of the formatting engine or adds new formatting functionality for statements not currently handled by the formatting engine.

That removes the cost of maintenance of 'esoteric' formatting options from the IDE team, and allows users to format their code however they want.

**Comment by DustinCampbell** [comment]
@tannergooding, that's out of scope at the moment, but is on the table for the future. The current formatting API is not suitable for making public, but it is already designed to be pluggable with additional rules.

**Comment by viceroypenguin** [comment]
> Obviously 😄
😄 

@CyrusNajmabadi, thanks for explaining your position here.  I don't think it's a bad position to have.  Management of a code base, especially such a large and complex one, is important to ensure that the team is ready and willing to maintain the ongoing support of the code.

I contributed because it was marked as 'Up For Grabs', implying that on some level there is interest from the Roslyn team for it to be implemented.  This may have been a quick decision, and now that it's implemented a larger conversation needs to be had.  

Regardless, I would say that if this is the position the team wants to take at some point, then when that decision is made, the Issues marked as Up For Grabs in the formatting space should be closed as 'Not Implementing'.

**Comment by CyrusNajmabadi** [comment]
It's just my position :smile:.  It may not be universal, or correct.  I'd like to hear what @DustinCampbell and @Pilchie think as they have good mindsets toward flexibility and the appropriate directions for our code.

**Comment by alrz** [comment]
resolved via https://github.com/dotnet/roslyn/issues/20292, close?

**Comment by tannergooding** [comment]
@viceroypenguin, I'm going to close this as I believe the original scenario was resolved by @CyrusNajmabadi in #20292. Feel free to reopen if you believe there is some scenario that is still uncovered.

Thank you for taking the time to put forth this contribution.

**Pull request closed without merging**